File: 067_SingleNumber.cpp
// Topic: Bit Manipulation
// Single Number: XOR all numbers to find unique element
// Time Complexity: O(n)
// Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

int singleNumber(vector<int>& nums)
{
    int x = 0;
    for (int v : nums) x ^= v;
    return x;
}

int main()
{
    cout << singleNumber((vector<int>{2, 2, 1})) << "\n";        // 1
    cout << singleNumber((vector<int>{4, 1, 2, 1, 2})) << "\n";  // 4
    return 0;
}

File: 067_SingleNumber.py
"""
Topic: Bit Manipulation
Problem: Single Number
Description: Find the element that appears once when others appear twice.
Time: O(n)
Space: O(1)
"""

from typing import List


def singleNumber(nums: List[int]) -> int:
    res = 0
    for x in nums:
        res ^= x
    return res


if __name__ == '__main__':
    print(singleNumber([2,2,1]))  # 1
    print(singleNumber([4,1,2,1,2]))  # 4
